Add REST implementation for Validator's `WaitForActivation` (Ethereum Protocol Fellowship)

nalepae opened the pull request:
### What type of PR is this?
Feature

### Related gitHub issue:
https://github.com/prysmaticlabs/prysm/issues/11580

### What does this PR do? Why is it needed?
This adds an implementation for the `WaitForActivation` validator client method by calling the Beacon Node REST API instead of the gRPC one.

### Comments for reviewers:

#### Minimal code mofication:
This PR is inspired by https://github.com/prysmaticlabs/prysm/pull/11654.
However, contrary to this one, I tried not to modify at all the file [validator/client/validator.go](https://github.com/prysmaticlabs/prysm/blob/develop/validator/client/validator.go).

**Pros:** It does not modify code currently used when gRPC is active.
**Cons:** The generated code is more convoluted, since I try to mock gRPC interface using REST.

Let me know which method (between the one in this PR and the one modifying [validator/client/validator.go](https://github.com/prysmaticlabs/prysm/blob/develop/validator/client/validator.go) like in https://github.com/prysmaticlabs/prysm/pull/11654) is best suited.

#### ValidatorsHttpResponse

I defined directly in the file `activation.go` the  data structures `ValidatorsHttpResponse`
It could be defined elsewhere in a specific directory if needed.

#### Validators status matching

Currently used gRPC validators status and beacon API validators status do not match 100%.
I had to define a mapping between them (`BeaconAPITogRPCValidatorStatus`)

#### Missing keys

When calling `eth/v1alpha1/validator/activation/stream` with a given set of keys, if a key is not known by the beacon node, the beacon node answers:
- `max(int)` validator Index
- `ValidatorStatus_UNKNOWN_STATUS` validator status

Asking `/eth/v1/beacon/states/head/validators?id=0x<aaa>` on beacon API does not return any item for `0x<aaa>` if this keys does not exists.
==> I emulated this behaviour in `waitForActivation` implementation.

#### Set object

`waitForActivation` uses set object.
Sets are not natively defined in golang, even if there is some libraries like [golang-set](https://github.com/deckarep/golang-set) which implements them.

I chose not to add an extra dependancy and used `map[string]struct{}` instead of set.
Let me know which choice is best suited.

PatriceVignola commented on validator/client/beacon-api/BUILD.bazel:
```diff
@@ -4,7 +4,10 @@ load("@prysm//tools/go:def.bzl", "go_library")
     name = "go_default_library",
-    srcs = ["beacon_api_validator_client.go"],
+    srcs = [
+        "activation",
```
Missing the `.go` extension

PatriceVignola commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,158 @@
+	}, nil
+}
+
+type ValidatorsHttpResponse struct {
```
We should be able to use `StateValidatorsResponseJson` from `beacon-chain/rpc/apimiddleware/structs.go`

nalepae replied:
Perfect! It's exactly what I was looking for!

PatriceVignola commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,158 @@
+	return &x.ValidatorActivationResponse, nil
+}
+
+var BeaconAPITogRPCValidatorStatus = map[string]ethpb.ValidatorStatus{
```
This makes it a public variable, and I guess we don't want that.

PatriceVignola commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,158 @@
+	"github.com/prysmaticlabs/prysm/v3/api/gateway/apimiddleware"
+	types "github.com/prysmaticlabs/prysm/v3/consensus-types/primitives"
+	ethpb "github.com/prysmaticlabs/prysm/v3/proto/prysm/v1alpha1"
+	"google.golang.org/grpc"
```
Is Bazel able to build with this new import? It's not in the `BUILD.bazel` file.

nalepae replied:
Actually I built it directly in VSCode.
There is a strong possibility it bypasses Bazel.
Let me check it.

nalepae replied:
Is there an easy way to automatically add dependancies?

[Documentation](https://docs.prylabs.network/docs/reading/bazel#do-i-have-to-edit-build-files-myself) states: 
```
bazel run //:gazelle -- fix
```
On develop branch, it shows:

```
INFO: Analyzed target //:gazelle (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
INFO: From BazelWorkspaceStatusAction stable-status.txt:
date: illegal option -- -
usage: date [-jnRu] [-r seconds|file] [-v[+|-]val[ymwdHMS]]
            [-I[date | hours | minutes | seconds]]
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
date: illegal option -- -
usage: date [-jnRu] [-r seconds|file] [-v[+|-]val[ymwdHMS]]
            [-I[date | hours | minutes | seconds]]
            [-f fmt date | [[[mm]dd]HH]MM[[cc]yy][.ss]] [+format]
Target //:gazelle up-to-date:
  bazel-bin/gazelle-runner.bash
  bazel-bin/gazelle
INFO: Elapsed time: 3.654s, Critical Path: 0.04s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
gazelle: /Users/manu/Kiln/prysm/beacon-chain/state/state-native/BUILD.bazel:5.12-42.7: could not merge expression
gazelle: /Users/manu/Kiln/prysm/cmd/validator/flags/BUILD.bazel:5.12-11.7: could not merge expression
gazelle: /Users/manu/Kiln/prysm/config/fieldparams/BUILD.bazel:5.12-8.7: could not merge expression
gazelle: /Users/manu/Kiln/prysm/validator/client/validator-client-factory/BUILD.bazel:18.12-21.7: could not merge expression
gazelle: /Users/manu/Kiln/prysm/validator/client/validator-client-factory/BUILD.bazel:24.12-30.7: could not merge expression
```

nalepae replied:
Added dependencies by hand

rkapka replied:
I think gazelle does not work here because of this:
```
# Gazelle will flag this for removal since it's only being used with the use_beacon_api flag, but it should be kept
# keep
```

nalepae commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+		return nil, errors.Errorf("error %d: %s", errorJson.Code, errorJson.Message)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
```
`err` not checked

prestonvanloon replied:
Yes, please handle this error.

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+	if resp.StatusCode != http.StatusOK {
+		errorJson := apimiddleware.DefaultErrorJson{}
+		err = json.NewDecoder(resp.Body).Decode(&errorJson)
+
```
```suggestion
```

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+	for _, data := range res.Data {
+		pubkey, err := hex.DecodeString(data.Validator.Pubkey[2:])
+
```
```suggestion
```

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	for _, data := range res.Data {
+		pubkey, err := hex.DecodeString(data.Validator.Pubkey[2:])
```
Please add a comment why we start at index `2`.

prestonvanloon replied:
Consider using go-ethereum's hexutil and passing the whole string.

https://pkg.go.dev/github.com/ethereum/go-ethereum/common/hexutil#Decode

This is better because your code will panic if pubkey length is less than 2 while go-ethereum's code has already handled such a use case.

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+	}
+
+	for missingPubKey := range missingPubKeys {
+		pubkeyBytes, err := hex.DecodeString(missingPubKey[2:])
```
Please add a comment why we start at index `2`.

prestonvanloon replied:
Same feedback here about using hexutil. 

https://pkg.go.dev/github.com/ethereum/go-ethereum/common/hexutil#Decode

rkapka reviewed (changes requested):
> Let me know which method (between the one in this PR and the one modifying [validator/client/validator.go](https://github.com/prysmaticlabs/prysm/blob/develop/validator/client/validator.go?rgh-link-date=2022-11-20T18%3A47%3A02Z) like in https://github.com/prysmaticlabs/prysm/pull/11654) is best suited.

I like your approach better because it's safer for existing code. We don't have to worry about regression bugs creeping in. The other approach is fine too, as long as it changes to existing code are minimal. e2e tests should help us find any bugs.

> I chose not to add an extra dependancy and used map[string]struct{} instead of set.
Let me know which choice is best suited.

I like what you have. It's clear and simple.

---

Two general comments:
1. The gRPC implementation polls statuses every slot until at least one is active: https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/beacon-chain/rpc/prysm/v1alpha1/validator/server.go#L98. Your code doesn't work like this. It simply returns validator statuses once.
2. Please add unit tests to `waitForActivation`. You should be able to use the same suggestion as https://github.com/prysmaticlabs/prysm/pull/11654#pullrequestreview-1188426496.

prestonvanloon commented on validator/client/beacon-api/BUILD.bazel:
```diff
@@ -4,12 +4,19 @@ load("@prysm//tools/go:def.bzl", "go_library")
 # keep
```
FYI you can add this here and it will fix the problem:

```
# gazelle:build_tag use_beacon_api
```

nalepae replied:
Fixed upstream

prestonvanloon commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+	statuses := []*ethpb.ValidatorActivationResponse_Status{}
+
+	for _, publicKey := range in.PublicKeys {
+		stringPubKey := "0x" + hex.EncodeToString(publicKey)
```
Prefer to use the go-ethereum hexutil. 

https://pkg.go.dev/github.com/ethereum/go-ethereum/common/hexutil#Encode

prestonvanloon commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+func (c beaconApiValidatorClient) waitForActivation(ctx context.Context, in *ethpb.ValidatorActivationRequest) (ethpb.BeaconNodeValidator_WaitForActivationClient, error) {
+	// Represents the set of keys in `in`
+	targetPubKeys := make(map[string]struct{})
```
Since you know the size of this map, help the runtime memory usage by specifying the exact size to pre-allocate.

```suggestion
	targetPubKeys := make(map[string]struct{}, len(in.PublicKeys))
```

prestonvanloon commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+	params := url.Values{}
+
+	for _, publicKey := range in.PublicKeys {
+		params.Add("id", fmt.Sprintf("0x%s", hex.EncodeToString(publicKey)))
+	}
```
You already did this string encoding once. Now you can use the `targetPubKeys` map keys to set up these params while reusing the encoding work that was already done previously. 

```suggestion
	for k := range targetPubKeys {
		params.Add("id", k)
	}
```

nalepae replied:
Yes, or I can iterate only once over `in.PublicKeys`.

prestonvanloon commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+	if resp.StatusCode != http.StatusOK {
+		errorJson := apimiddleware.DefaultErrorJson{}
+		err = json.NewDecoder(resp.Body).Decode(&errorJson)
+
+		if err != nil {
+			return nil, err
+		}
```
Two suggestions:

- Use `if err := fn(); err != nil` for a more idiomatic go approach
- Wrap the error to help users understand what went wrong

```suggestion
		if err := json.NewDecoder(resp.Body).Decode(&errorJson); err != nil {
			return nil, errors.Wrapf(err, "received status code %d and failed to decode response body error json", resp.StatusCode)
		}
```

prestonvanloon commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,162 @@
+
+	var res ValidatorsHttpResponse
+	if err := json.Unmarshal(body, &res); err != nil {
+		return nil, err
```
Consider wrapping this error as well. It can really help with debugging.

```suggestion
		return nil, errors.Wrap(err, "failed to decode validators http response")
```

nalepae replied:
Should I wrap all occurrences of this kind of code
```go
if err != nil {
    return nil, err
}
```

with `errors.Wrap` and?

prestonvanloon reviewed (changes requested):
Please add tests

nalepae commented:
@rkapka @prestonvanloon ,

I moved a lot of the logic into `Recv` and added tests.

With gRPC, the `wait` part is done on Beacon side.
With REST, the `wait` part is done on Validator side.

Note the `wait` part is not covered by test, since it adds a long period (12 seconds) of testing time.
If needed, adding the `wait` line to coverage is very easily doable. (We just need to call `Recv` 2 times.)

nalepae commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,153 @@
+	firstRecvCall bool
+}
+
+func (x *beaconNodeValidatorWaitForActivationClient) Recv() (*ethpb.ValidatorActivationResponse, error) {
```
My understanding is the code you quote:

https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/beacon-chain/rpc/prysm/v1alpha1/validator/server.go#L95-L118

is running on server side (Beacon Node) when using gRPC and not on client side (Validator Client).

> But how does the caller know when to stop looping?

The validator client actually runs a `for` loop here: 
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212

It calls `Recv` on line 169.
- If no validator is active, it fails test on line 203 then goes to `continue` on line 206 (then goes back to the beginning of the loop).
- If at least one validator is active, it successes test on line 203 then goes to `break` line 209 then exits the `for` loop.

So, if the validator starts with no active key, it will continue to run the for loop.
Once at least once key is active, it will exit the for loop.

Note: Each call of `Recv` asks the BN for latest keys statuses.

-----------
With gRPC, there is actually 2 `for` loops running:
The first one this one, in the Beacon Node:
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/beacon-chain/rpc/prysm/v1alpha1/validator/server.go#L95-L118

The second one is this one, in the Validator Client:
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212


Using REST, there is no more `for` loop on Beacon Node.

==> I can incorporate into the `Recv` function  a `for+select+channel` with a `time.After` clause every 12 seconds which will loop and print some logs every slot until at least one validator is active, and which will return the list of statuses if one of them is active.

It will solve the "drift" problem, but will add some redundancy with the code here
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212

which already loops until one validator is active.

(As I said earlier, I'm quite new at golang, so I may miss a very obvious point related to the language!)

nalepae replied:
Let me come back with something more convincing :)

nalepae replied:
I don't see the functional need to emulate on client side the for loop currently present on server side with gRPC.
I have the impression (at least, when I test it) that the behaviour is correct, except for the drift issue.
I can mitigate this drift issue by using a `for+select+channel` instead of the `wait`.

nalepae commented:
- If no validator is in active state, `Recv` method waits for the time needed to reach 1 slot duration (12 seconds) to pass between the previous call and the current call => No time drift anymore.
- The `Recv` method always returns `ValidatorActivationResponse` containing the latest set of statuses.
- If the context is canceled during the wait of `Recv` method, the wait stops early and the method returns.
- Once at least one validator is active, the corresponding status is returned from the `Recv` method, and the following for loop
https://github.com/prysmaticlabs/prysm/blob/d58d3f2c57878e4ba06411e77ee0c1a8ffc9d2e3/validator/client/wait_for_activation.go#L163-L212

exits with lines 203, 204 and 209.

Difference with gRPC version:
- No more `for` loop on server (Beacon) side.
- With gRPC version, the beacon sends statuses to the stream exactly once per slot, even if the validator is not asking for it. When the validator call the `Recv` method, it does not receive the latest statutes, but the first not de-queued from the stream. With REST version, when calling `Recv`, the validator received the latest statuses, at a rate of 1 statuses every slot maximum.

Let me know if there is a functional requirement I miss!

(@rkapka)

rkapka reviewed (commented):
I personally dislike the fact that a type called `beaconNodeValidatorWaitForActivationClient` doesn't keep streaming by itself when there are no active validators. Also because `x.lastRecvTime` is preserved between calls, two different parts of code invoking this function will interfere with each other.

nisdas commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,166 @@
+			params.Add("id", stringPubKey)
+		}
+
+		resp, err := x.beaconApiValidatorClient.httpClient.Get(x.beaconApiValidatorClient.url + "/eth/v1/beacon/states/head/validators?" + params.Encode())
```
Can we better represent this http request, hardcoding the url this way is going to be bug prone.

nalepae replied:
Created a `buildURL` function, so the corresponding code now looks like:

```go
	params := neturl.Values{}

	for pubkey := range pubkeys {
		params.Add("id", pubkey)
	}

	url := buildURL(
		x.beaconApiValidatorClient.url,
		"eth/v1/beacon/states/head/validators",
		params,
	)

	resp, err := x.beaconApiValidatorClient.httpClient.Get(url)
```

I hope it is cleaner.

nisdas commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,166 @@
+			}
+		}()
+
+		if resp.StatusCode != http.StatusOK {
```
Http request and responses should be abstracted into their own method and away from the main `WaitForActivation` loop

nalepae replied:
Created dedicated function for request/response `getStateValidators`.

nalepae commented:
> I personally dislike the fact that a type called beaconNodeValidatorWaitForActivationClient doesn't keep streaming by itself when there are no active validators.

I agree with the concern about naming.
General question: Is there a functional need to have a stream when there is no active validators? What do we functionnaly miss if there is not such a stream?

I'm not against implementing such a behaviour. However, with gRPC Beacon API, the Beacon Node itself is in charge of fooding such a stream. With REST Beacon API, it is not the case any more.

I see two solutions:
- We could either re-implement such a stream on Beacon Node side. In this case, the communication between Validator and Beacon won't be done via "official" Beacon API.
- We could emulate this behaviour in Validator Client. In this case, I don't see a way to achieve this emulation without modifying source code from Validator Client which is already used by current gRPC implementation. (Less safe for existing code.) But maybe there is a way I don't see! 

> Also because x.lastRecvTime is preserved between calls, two different parts of code invoking this function will interfere with each other.

Yes, if we retrieve a `beaconNodeValidatorWaitForActivationClient` object via the method `waitForActivation`, and if 2 different parts of code call the `Recv` method on the same previously retrieved  `beaconNodeValidatorWaitForActivationClient` object, then the 2 different parts will interfere with each other.

But if two different parts of the code retrieve two different `beaconNodeValidatorWaitForActivationClient` objects via the method `waitForActivation`, there is no interference.

Unless I'm mistaken, we already have kind of the same interfering behaviour with the current gRPC implementation: The call to `WaitForActivation` will open a stream and will return a object. The call of `Recv` method on this returned object will un-queue objects previously queued by the beacon node. So if 2 different parts of code call the `Recv` method on the same previously retrieved  `beaconNodeValidatorWaitForActivationClient` object, then the 2 different parts will interfere with each other.

rkapka reviewed (approved):
Thanks a lot for this! @nisdas left a few comments too.

nalepae commented:
I just fixed last @nisdas comments.

PatriceVignola commented on validator/client/beacon-api/activation_test.go:
```diff
@@ -0,0 +1,197 @@
+//go:build use_beacon_api
```
The case where a non-200 status is returned doesn't seem to be here. it would be good to test that scenario with both the valid error json case, and the badly formatted error json case.

nalepae replied:
All cases (200, 400, 404, 500, others) are now checked.
About timeout, I don't want the test to last 12+ secs.

==> I Check `<- time.After(waitDuration` is blocking when `Recv` is called two times in a raw, and I check the computation of `waitDuration` (moved in a new dedicated function) is done correctly.

(Fixed a bug on this computation)

PatriceVignola commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,185 @@
+		}
+
+		for _, data := range stateValidators.Data {
+			pubkey, err := hexutil.Decode(data.Validator.PublicKey)
```
This will panic if data.Validator is nil

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+		for _, data := range stateValidators.Data {
+			pubkey, err := hexutil.Decode(data.Validator.PublicKey)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to read validator public key")
```
```suggestion
				return nil, errors.Wrap(err, "failed to decode validator public key")
```

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+
+			index, err := strconv.ParseUint(data.Index, 10, 64)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to read validator index")
```
```suggestion
				return nil, errors.Wrap(err, "failed to parse validator index")
```

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+	if resp.StatusCode != http.StatusOK {
+		errorJson := &apimiddleware.DefaultErrorJson{}
+		if err := json.NewDecoder(resp.Body).Decode(&errorJson); err != nil {
+			return nil, errors.Wrap(err, "failed to decode response body validators error json")
```
```suggestion
			return nil, errors.Wrap(err, "failed to decode error json")
```
I think this is too cryptic

nalepae replied:
Code deleted

rkapka commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+
+	validatorsJson := &rcpmiddleware.StateValidatorsResponseJson{}
+	if err := json.NewDecoder(resp.Body).Decode(&validatorsJson); err != nil {
+		return nil, errors.Wrap(err, "failed to decode response body validators json")
```
```suggestion
		return nil, errors.Wrap(err, "failed to decode response json")
```
I think this is too cryptic

nalepae replied:
Code deleted

rkapka reviewed (commented):
Looks good, just a few minor comments

nisdas commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+	}, nil
+}
+
+type beaconNodeValidatorWaitForActivationClient struct {
```
Can we make the naming less verbose here ?

nalepae replied:
Renamed into `waitForActivationClient`

nisdas commented on validator/client/beacon-api/activation.go:
```diff
@@ -0,0 +1,184 @@
+	return nextRecvTime.Sub(now), nextRecvTime
+}
+
+func (x *beaconNodeValidatorWaitForActivationClient) Recv() (*ethpb.ValidatorActivationResponse, error) {
```
A better var name here rather than `x` . Since its a client `c` or `b` would be more readable

nalepae replied:
Renamed in `c`

nisdas reviewed (commented):
I had some cosmetic comments, but other than those it looks good to me if @rkapka has no
other concerns.

nalepae commented:
In this implementation I need the following duplication:

```go
stringTargetPubKeysToPubKeys := make(map[string][]byte, len(c.ValidatorActivationRequest.PublicKeys))
stringTargetPubKeys := make([]string, len(c.ValidatorActivationRequest.PublicKeys))
```
because :
- the function `getStateValidators` takes a list of strings as an input, and
- I also need to check if a given key is present in target pub keys (`(O(1)` lookup for a slice, `O(1)` for a map)

I guess the memory duplication is acceptable (even in the very unlikely case where a single validator get 500k validators keys...).

If not acceptable, I can propose 2 alternative solutions.

**First alternative solution:**
Go does not provide built-in function to convert a keys of a map to a slice, but there is one in experimental package.

```go
import (
    "golang.org/x/exp/maps"
)

...

stateValidators, err := c.beaconApiValidatorClient.getStateValidators(maps.Keys(stringTargetPubKeysToPubKeys))
```

**Second alternative solution:**
I can modify the `getStateValidators` for it to accept a `map[string][]byte`) instead of a `[]string`.

nalepae commented:
@prestonvanloon, are you able to unblock the merge if changes you requested are OK?
